cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -422,6 +426,93 @@ impl YaccParser {
         }
     }
 
+    // Handle parse_param declarations of the form:
+    // %parse_param <'a>(x: u32, y : (u32, u32))
+    fn parse_param(&mut self, mut i: usize) -> YaccResult<usize> {
+        i = self.parse_ws(i, false)?;
+        // First gobble up all of the '<' lifetime ',' ... '>
+        if let Some(mut j) = self.lookahead_is("<", i) {
+            let mut k = j;
+            let mut lifetimes = HashSet::new();
+            let mut add_lifetime = |k, c: char| {
+                let s = self.src[j..k].trim_start().trim_end().to_string();
+                lifetimes.insert(s);
+                j = k + c.len_utf8();
+                j
+            };
+
+            while k < self.src.len() {
+                let c = self.src[k..].chars().next().unwrap();
+                match c {
+                    '\n' | '\r' => return Err(self.mk_error(YaccParserErrorKind::ReachedEOL, k)),
+                    ',' => {
+                        k = add_lifetime(k, ',');
+                        continue;
+                    }
+                    '>' => {
+                        k = add_lifetime(k, '>');
+                        break;
+                    }
+                    _ => k += c.len_utf8(),
+                }
+            }
+            self.ast.parse_param_lifetimes = Some(lifetimes);
+            i = k;
+        }
+
+        // Next, the '(' pattern : type, ... ')'
+        i = self.parse_ws(i, false)?;
+        if self.lookahead_is("(", i).is_some() {
+            let mut j = i;
+            let mut bindings: Vec<(String, String)> = Vec::new();
+            while j < self.src.len() && self.lookahead_is(")", j).is_none() {
+                let c = self.src[j..].chars().next().unwrap();
+                j += c.len_utf8();
+
+                // Some binding name, or pattern.
+                let k = self.parse_ws(j, false)?;
```

Comment:
Since we don't reference `j` afterwards, I think this can be just `j = self.parse_ws(...)`.

Reply:
Yeah, what I wouldn't give for `(k, let foo) = ...`.

Reply:
Agreed -- it does lead to some ugly code :/